tests/contrib/celery/test_utils.py

```diff
@@ -36,13 +44,153 @@ def test_meta_from_context(self):
         ok_(metas.get('custom_meta', None) is None)
 
-    def test_meta_from_context_empty_keys(self):
+    def test_tags_from_context_empty_keys(self):
         # it should not extract empty keys
         context = {
             'correlation_id': None,
+            'exchange': '',
             'timelimit': (None, None),
             'retries': 0,
         }
 
-        metas = meta_from_context(context)
-        eq_({}, metas)
+        tags = tags_from_context(context)
+        eq_({}, tags)
+        # edge case: `timelimit` can also be a list of None values
+        context = {
+            'timelimit': [None, None],
+        }
+
+        tags = tags_from_context(context)
+        eq_({}, tags)
+
+    def test_span_propagation(self):
+        # ensure spans getter and setter works properly
+        @self.app.task
+        def fn_task():
+            return 42
+
+        # propagate and retrieve a Span
+        task_id = '7c6731af-9533-40c3-83a9-25b58f0d837f'
+        span_before = self.tracer.trace('celery.run')
+        propagate_span(fn_task, task_id, span_before)
+        span_after = retrieve_span(fn_task, task_id)
+        ok_(span_before is span_after)
+
+    def test_span_delete(self):
+        # ensure the helper removes properly a propagated Span
+        @self.app.task
+        def fn_task():
+            return 42
+
+        # propagate a Span
+        task_id = '7c6731af-9533-40c3-83a9-25b58f0d837f'
+        span = self.tracer.trace('celery.run')
+        propagate_span(fn_task, task_id, span)
+        # delete the Span
+        weak_dict = getattr(fn_task, '__dd_task_span')
+        remove_span(fn_task, task_id)
+        ok_(weak_dict.get(task_id) is None)
+
+    def test_span_delete_empty(self):
+        # ensure the helper works even if the Task doesn't have
+        # a propagation
+        @self.app.task
+        def fn_task():
+            return 42
+
+        # delete the Span
+        exception = None
+        task_id = '7c6731af-9533-40c3-83a9-25b58f0d837f'
+        try:
+            remove_span(fn_task, task_id)
+        except Exception as e:
+            exception = e
+        ok_(exception is None)
+
+    def test_memory_leak_safety(self):
+        # Spans are shared between signals using a Dictionary (task_id -> span).
+        # This test ensures the GC correctly cleans finished spans. If this test
+        # fails a memory leak will happen for sure.
+        @self.app.task
+        def fn_task():
+            return 42
+
+        # propagate and finish a Span for `fn_task`
+        task_id = '7c6731af-9533-40c3-83a9-25b58f0d837f'
+        propagate_span(fn_task, task_id, self.tracer.trace('celery.run'))
+        weak_dict = getattr(fn_task, '__dd_task_span')
+        ok_(weak_dict.get(task_id))
+        # flush data and force the GC
+        weak_dict.get(task_id).finish()
+        self.tracer.writer.pop()
+        self.tracer.writer.pop_traces()
+        gc.collect()
```

Comment:
because we have weak references, this means the element is eligible for garbage collection at any time after the `Span` is finished. It's not garbage collected if we still have references of the `Span`. This is needed because it's possible the signal mechanism is not called properly from Celery (unexpected behaviors / unhandled executions) and so even if a Span is not finished it can be garbage collected if the Tracer releases the reference to the `Span`.

We may need to check if we have to connect to other signals to be sure Spans are always finished.